spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/RNN.scala

```diff
@@ -347,32 +393,208 @@ class RNN(
+
+    weightForBackwardMemoryData = realWei
+    reorderManager.register(weight.heapData, realWei)
+    weightForBackward = reorderManager
+      .infer(Array(weight.heapData), Array(weightForBackwardMemoryData), weight.dense)
+      .asInstanceOf[DnnTensor[Float]]
+
+    weightIterForBackwardMemoryData = realWei_iter
+    reorderManager.register(weight_i.heapData, realWei_iter)
+    weightIterForBackward = reorderManager
+      .infer(Array(weight_i.heapData), Array(weightIterForBackwardMemoryData), weight_i.dense)
+      .asInstanceOf[DnnTensor[Float]]
+
+    gradsrc_i.setMemoryData(realDiffSrc_iter, HeapData(gradsrc_i.size(), Memory.Format.ldsnc),
+      runtime)
+    graddst_i.setMemoryData(realDiffDst_iter, HeapData(graddst_i.size(), Memory.Format.ldsnc),
+      runtime)
+    gradWeight.setMemoryData(realDiffWei, HeapData(gradWeight.size(), Memory.Format.ldigo), runtime)
+    gradWeight_i.setMemoryData(realDiffWei_iter, HeapData(gradWeight_i.size(), Memory.Format.ldigo),
+      runtime)
+    gradBias.setMemoryData(realDiffBias, HeapData(gradBias.size(), Memory.Format.ldgo), runtime)
+
+    gradsrc_i.zero()
+    graddst_i.zero()
+    gradWeight.zero()
+    gradWeight_i.zero()
+    gradBias.zero()
+
+    val srcs = Array(realSrc.getPrimitive(runtime), realSrc_iter.getPrimitive(runtime),
+      realWei.getPrimitive(runtime), realWei_iter.getPrimitive(runtime),
+      realBias.getPrimitive(runtime), realDst.getPrimitive(runtime),
+      realDst_iter.getPrimitive(runtime), realDiffDst.getPrimitive(runtime),
+      realDiffDst_iter.getPrimitive(runtime), workSpaceFormat.getPrimitive(runtime))
+    val indexes = Array.fill(srcs.length)(0)
+
+    val dsts = Array(realDiffSrc.getPrimitive(runtime), realDiffSrc_iter.getPrimitive(runtime),
+      realDiffWei.getPrimitive(runtime), realDiffWei_iter.getPrimitive(runtime),
+      realDiffBias.getPrimitive(runtime)
+    )
+
+    val primitive = MklDnn.PrimitiveCreate2(bwdPD, srcs, indexes, srcs.length,
+      dsts, dsts.length)
+
+    updateGradInputMemoryPrimitives = srcs ++ dsts
+    updateGradInputPrimitives = Array(primitive)
+    gradInput = initTensor(realDiffSrc)
+
+    _gradInputFormats = Array(realDiffSrc)
+    _gradOutputFormats = Array(realDiffDst)
+    (_gradOutputFormats, _gradInputFormats)
+  }
+
+  override def updateGradInput(input: Activity, gradOutput: Activity): Activity = {
+    if (updateGradInputTensors == null) {
+      val buffer = new ArrayBuffer[Tensor[Float]]()
+      buffer.append(input.asInstanceOf[Tensor[Float]]) // 0
+      buffer.append(src_i.native) // 1
+      buffer.append(weightForBackward) // 2
+      buffer.append(weightIterForBackward) // 3
+      buffer.append(bias.native) // 4
+      buffer.append(output.asInstanceOf[Tensor[Float]]) // 5
+      buffer.append(dst_i.native) // 6
+      buffer.append(gradOutput.asInstanceOf[Tensor[Float]]) // 7
+      buffer.append(graddst_i.native) // 8
+      buffer.append(workSpace) // TODO 9
```

Comment:
what does this `todo` mean?

Reply:
Forgot to remove them before push. Deleted